### PR TITLE
Grpc 19871/unary unary client aio implement timeout

### PR DESCRIPTION
**Pull request by rmariano**
Implement timeout for the async unary_unary call

Allow passing the ``timeout`` parameter in the asynchronous version of  the unary_unary call, and use it accordingly.

Maintains the same interface as the synchronous version.

Testing results
```
GRPC_ENABLE_FORK_SUPPORT=0 python ./src/python/grpcio_tests/setup.py test_aio
running test_aio
Running       _sanity._sanity_test.AioSanityTest.testTestsJsonUpToDate
Testing gRPC Python...
SUCCESS       _sanity._sanity_test.AioSanityTest.testTestsJsonUpToDate
Running       unit.channel_test.TestChannel.test_async_context
SUCCESS       unit.channel_test.TestChannel.test_async_context
Running       unit.channel_test.TestChannel.test_unary_call_times_out
SUCCESS       unit.channel_test.TestChannel.test_unary_call_times_out
Running       unit.channel_test.TestChannel.test_unary_unary
SUCCESS       unit.channel_test.TestChannel.test_unary_unary
Running       unit.init_test.TestInsecureChannel.test_insecure_channel
SUCCESS       unit.init_test.TestInsecureChannel.test_insecure_channel
5 tests finished:
	5 successful
	0 unsuccessful
	0 skipped
	0 expected failures
	0 unexpected successes
Interrupted Tests:
	[]

Errors/Failures:

Unexpected successes: []
```

Fixes https://github.com/grpc/grpc/issues/19871

### CODE REVIEW

File: src/python/grpcio/grpc/_cython/_cygrpc/aio/call.pyx.pxi [hunk 1]
```diff
@@ -146,4 +168,15 @@ cdef class _AioCall:
             grpc_call_unref(call)
             gpr_free(ops)
 
-        return receive_message_operation.message()
+        if receive_status_on_client_operation.code() == GRPC_STATUS_OK:
+            return receive_message_operation.message()
+
+        class _RpcError(_RpcErrorBase, grpc.RpcError):
```
Comment (argaen) [inline src/python/grpcio/grpc/_cython/_cygrpc/aio/call.pyx.pxi]: is there a requirement on declaring this class on every `unary_unary` call? Can't we extract it outside?

File: src/python/grpcio_tests/tests_aio/unit/channel_test.py [hunk 2]
```diff
@@ -39,16 +42,34 @@ def test_async_context(self):
     def test_unary_unary(self):
 
         async def coro():
-            channel = aio.insecure_channel(self.server_target)
-            hi = channel.unary_unary(
-                '/grpc.testing.TestService/UnaryCall',
-                request_serializer=messages_pb2.SimpleRequest.SerializeToString,
-                response_deserializer=messages_pb2.SimpleResponse.FromString)
-            response = await hi(messages_pb2.SimpleRequest())
+            async with aio.insecure_channel(self.server_target) as channel:
```
Comment (argaen) [inline src/python/grpcio_tests/tests_aio/unit/channel_test.py]: by doing this you remove the test that checks the explicit call of `aio.insecure_channel` without a context manager. Note that above there is a test for context manager.

I would at least leave one test that checks for the explicit call

File: src/python/grpcio_tests/tests_aio/unit/sync_server.py [hunk 3]
```diff
@@ -26,9 +26,15 @@
 # Change for an asynchronous server version once it's implemented.
 class TestServiceServicer(test_pb2_grpc.TestServiceServicer):
 
+    CALL_DELAY = 0.2
+
     def UnaryCall(self, request, context):
         return messages_pb2.SimpleResponse()
 
+    def EmptyCall(self, request, context):
```
Comment (argaen) [inline src/python/grpcio_tests/tests_aio/unit/sync_server.py]: I think this discussion was somewhere else already but I don't find it. Why can't we call this `SleepCall` or `DelayCall` or something like that?

Comment (rmariano) [reply src/python/grpcio_tests/tests_aio/unit/sync_server.py]: This server's implementation depends on the definition of the protobuf, so the methods must be called as per that definition. Calling it something else (``TimeoutCall``, ``DelayedCall``, etc.) requires us to come up with a new protobuf, and a new server object that implements it. So in order to reuse the sever we already have, I implemented another of the mehtods that was left free.

File: src/python/grpcio_tests/tests_aio/unit/channel_test.py (hunk 2 above)
Comment (rmariano) [reply src/python/grpcio_tests/tests_aio/unit/channel_test.py]: Reverted 👍

File: src/python/grpcio/grpc/_cython/_cygrpc/aio/call.pyx.pxi (hunk 1 above)
Comment (rmariano) [reply src/python/grpcio/grpc/_cython/_cygrpc/aio/call.pyx.pxi]: If this is declared in the main file, it doesn't compile property because it raises ``ImportError: 'grpc' does not have attribute RpcError`` (becaues it's a cython file calling a Python one), so it was done this way to have it inside the compiled code, so at running time the dependency will be availalble.

TL;DR the main constraing is ``grpc.RpcError`` being a Python class needed from a Cython one.

**Comment by pfreixes** [comment]
PTAL @lidizheng & @gnossen

File: src/python/grpcio/grpc/experimental/aio/_channel.py [hunk 4]
```diff
@@ -20,12 +21,18 @@
...
-                 response_deserializer):
+    def __init__(self, channel: cygrpc.AioChannel, method, request_serializer,
+                 response_deserializer) -> None:
         self._channel = channel
         self._method = method
         self._request_serializer = request_serializer
         self._response_deserializer = response_deserializer
+        self._loop = asyncio.get_event_loop()
+
+    def _timeout_for_deadline(self, timeout=None):
```
Comment (gnossen) [inline src/python/grpcio/grpc/experimental/aio/_channel.py]: Is there a reason why timeout is a keyword argument when it's basically the only argument?

File: src/python/grpcio_tests/tests_aio/unit/channel_test.py [hunk 5]
```diff
@@ -52,6 +55,26 @@ def test_unary_unary(self):
...
+    def test_unary_call_times_out(self):
+        async def coro():
+            async with aio.insecure_channel(self.server_target) as channel:
+                empty_call_with_sleep = channel.unary_unary(
+                    "/grpc.testing.TestService/EmptyCall",
+                    request_serializer=messages_pb2.SimpleRequest.SerializeToString,
+                    response_deserializer=messages_pb2.SimpleResponse.FromString,
+                )
+                timeout = TestServiceServicer.CALL_DELAY / 2
+                with self.assertRaises(grpc.RpcError) as exception_context:
```
Comment (gnossen) [inline src/python/grpcio_tests/tests_aio/unit/channel_test.py]: I could see this test potentially being flaky depending on how accurate the various timers involved in this RPC are and scheduling of the threads and processes involved. What sort of flake rate do we see when running several thousand instances of this test in parallel?

File: src/python/grpcio/grpc/_cython/_cygrpc/aio/channel.pyx.pxi [hunk 6]
```diff
@@ -18,13 +18,13 @@ cdef class AioChannel:
...
-        class_name = self.__class__.__name__ 
+        class_name = self.__class__.__name__
         id_ = id(self)
         return f"<{class_name} {id_}>"
 
     def close(self):
         grpc_channel_destroy(self.channel)
 
-    async def unary_unary(self, method, request):
+    async def unary_unary(self, method, request, timeout=None):
```
Comment (gnossen) [inline src/python/grpcio/grpc/_cython/_cygrpc/aio/channel.pyx.pxi]: Our pattern in the past has been to use keyword arguments on our public API surface for breathing room for API additions, but for internal APIs such as this, we tend to only use positional arguments, since we're completely in control of the call sites as well as the function signature.

File: src/python/grpcio/grpc/_cython/_cygrpc/aio/call.pyx.pxi [hunk 7]
```diff
@@ -146,4 +168,15 @@ cdef class _AioCall:
             grpc_call_unref(call)
             gpr_free(ops)
 
-        return receive_message_operation.message()
+        if receive_status_on_client_operation.code() == GRPC_STATUS_OK:
```
[the remark this answers is not in the document]
Comment (rmariano) [reply src/python/grpcio/grpc/_cython/_cygrpc/aio/call.pyx.pxi]: I would like to reuse that enum, but ``grpc.StatusCode.OK`` it's an enum whose value is ``(0, 'ok')``, and the result of the ``.code()`` method is just the status code number (``0``), so they don't match. Maybe I would unpack just the status code from the enum, and using that to compare.

I've noticed in the current check of that condition, the ``state`` object being compared is of a different kind, whcih is not available in the message operation objects we use here. Would it be worth re-assempling the tuple to use that on the check?

[the remark this answers is not in the document]
Comment (rmariano) [reply src/python/grpcio/grpc/_cython/_cygrpc/aio/call.pyx.pxi]: In the end, I resolved by using ``StatusCode.ok`` from ``cygrpc`` 👍

File: src/python/grpcio_tests/tests_aio/unit/channel_test.py (hunk 5 above)
Comment (rmariano) [reply src/python/grpcio_tests/tests_aio/unit/channel_test.py]: Now that the code uses ``SHORT_TIMEOUT`` from ``tests_constants.py``, it has a larger value, so I would estimate that reduces the flakyness of the test. Would you agree with this assessment?

**Comment by rmariano** [comment]
Hi @lidizheng , @gnossen Thanks for the feedback! I have addressed your comments.
Please let me know what you think now 😃

File: src/python/grpcio_tests/tests_aio/unit/init_test.py [hunk 8]
```diff
@@ -15,10 +15,33 @@
...
+        aio_rpc_error = grpc.AioRpcError("initial metadata", 0, "details", "trailing metadata")
+
+        self.assertEqual(aio_rpc_error.__class__.__name__, "AioRpcError")
+        self.assertIsInstance(aio_rpc_error, grpc.RpcError)
+
+    def test_class_singleton(self):
+        first_aio_rpc_error = grpc.AioRpcError("initial metadata", 0, "details", "trailing metadata")
+        second_aio_rpc_error = grpc.AioRpcError("initial metadata", 0, "details", "trailing metadata")
+
+        self.assertEqual(id(first_aio_rpc_error.__class__), id(second_aio_rpc_error.__class__))
```
Comment (gnossen) [inline src/python/grpcio_tests/tests_aio/unit/init_test.py]: Nice test addition! :smiley: 

There's a built-in assertion for this:
```python
self.assertIs(first_aio_rpc_error.__class__, second_aio_rpc_error.__class__)
```

File: src/python/grpcio_tests/tests_aio/unit/channel_test.py (hunk 5 above)
Comment (gnossen) [reply src/python/grpcio_tests/tests_aio/unit/channel_test.py]: :man_shrugging: Your guess is as good as mine. I was hoping for a measurement.

In fact, I would actually prefer a timeout as short as possible and some mechanism that guarantees the server doesn't respond until after the client times out. Could you have the server sleep forever instead of for a finite period of time?

File: src/python/grpcio/grpc/__init__.py [hunk 9]
```diff
@@ -299,6 +299,27 @@ class RpcError(Exception):
     """Raised by the gRPC library to indicate non-OK-status RPC termination."""
 
 
+class AioRpcError(RpcError):
```
Comment (gnossen) [inline src/python/grpcio/grpc/__init__.py]: If this goes directly into `__init__.py`, it signals inclusion into our stable public API. Can we somehow make this live under the `experimental` module?

**Comment by rmariano** [comment]
Hi @gnossen @lidizheng I've updated the PR addressing your comments (thanks for the feedback!). Could you please take another look?

File: src/python/grpcio/grpc/__init__.py [hunk 10]
```diff
@@ -1963,6 +1963,7 @@ class Compression(enum.IntEnum):
     'StatusCode',
     'Status',
     'RpcError',
+    'AioRpcError',
```
Comment (lidizheng) [inline src/python/grpcio/grpc/__init__.py]: I think we shouldn't touch our API surface now. `AioRpcError` should belong to the init file of experimental/aio instead of the main init file.

File: src/python/grpcio/grpc/_cython/_cygrpc/aio/rpc_error.pxd.pxi [hunk 11]
```diff
@@ -0,0 +1,27 @@
...
+# limitations under the License.
+"""Exceptions for the aio version of the RPC calls."""
+
+
+cdef class _AioRpcError(Exception):
+    cdef readonly:
+        object _initial_metadata
+        int _code
+        str _details
+        object _trailing_metadata
```
Comment (lidizheng) [inline src/python/grpcio/grpc/_cython/_cygrpc/aio/rpc_error.pxd.pxi]: nit: Both `object` can be replaced by `tuple`. It gives Cython chance to optimize a bit. Don't forget to change `__cinit__` in `pyx` file as well.

File: src/python/grpcio/grpc/experimental/aio/_channel.py [hunk 12]
```diff
@@ -20,24 +21,28 @@
 
 class UnaryUnaryMultiCallable(aio.UnaryUnaryMultiCallable):
 
-    def __init__(self, channel, method, request_serializer,
-                 response_deserializer):
+    def __init__(self, channel: cygrpc.AioChannel, method, request_serializer,
+                 response_deserializer) -> None:
```
Comment (lidizheng) [inline src/python/grpcio/grpc/experimental/aio/_channel.py]: Since you just added type to `channel`, why not add types to all arguments? It would probably be simpler to predefine those types, since they might be reused for other `Callable` class.

File: src/python/grpcio/grpc/experimental/aio/_channel.py [hunk 13]
```diff
@@ -20,24 +21,28 @@
...
         self._channel = channel
         self._method = method
         self._request_serializer = request_serializer
         self._response_deserializer = response_deserializer
+        self._loop = asyncio.get_event_loop()
+
+    def _timeout_for_deadline(self, timeout):
```
Comment (lidizheng) [inline src/python/grpcio/grpc/experimental/aio/_channel.py]: How about s/_timeout_for_deadline/_timeout_to_deadline/

File: src/python/grpcio_tests/tests_aio/unit/channel_test.py (hunk 5 above)
Comment (lidizheng) [reply src/python/grpcio_tests/tests_aio/unit/channel_test.py]: Can we add a TODO here? Since currently the client and server are not in a process, it would be hard to synchronize between them. But eventually they should be tested in the same process. Once the asyncio server is available, we can remove the sleep mechanism.

File: src/python/grpcio_tests/tests_aio/unit/channel_test.py [hunk 14]
```diff
@@ -52,6 +55,26 @@ def test_unary_unary(self):
 
         self.loop.run_until_complete(coro())
 
+    def test_unary_call_times_out(self):
```
Comment (lidizheng) [inline src/python/grpcio_tests/tests_aio/unit/channel_test.py]: It would be helpful to add a sunny day test case, just specify a timeout, and make sure it doesn't break the normal RPC.

**Comment by lidizheng** [comment]
@rmariano It seems the sanity check has failed. Can you `yapf` and `pylint` the code?
```
bash ./tools/distrib/yapf_code.sh
bash ./tools/distrib/pylint_code.sh
```

**Comment by rmariano** [comment]
@lidizheng Do you have an ETC for merging? Thanks

**Comment by lidizheng** [comment]
@rmariano I not familiar with the acronym ETC. The code is reviewed and fine to check in. What we need is to get the required tests to be passed. I can help if un-related tests broke, or some minor formatting issue.

**Comment by rmariano** [comment]
@lidizheng sorry I meant "estimated to completion".
I just fixed the linting issues, and the rest of the issues, weren't marked as required, so let me know if there is anything else that I need to take care of.

**Comment by lidizheng** [comment]
@rmariano No specific deadline, but since it is finished I hope it can be checked in quickly.

**Comment by lidizheng** [comment]
Known failures: https://github.com/grpc/grpc/issues/20519 https://github.com/grpc/grpc/issues/20502

Thank you for fixing the Sanity Test complains.

**Comment by nicolasnoble** [comment]
This is causing troubles with our internal version - reverting for now, @lidizheng will take the lead in re-integrating it.
